File: apps/api/src/main/java/com/example/demo/service/TaskHistoryService.java
package com.example.demo.service;

import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Objects;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.example.demo.model.Task;
import com.example.demo.model.TaskHistory;
import com.example.demo.model.TaskHistory.HistoryAction;
import com.example.demo.model.User;
import com.example.demo.repository.TaskHistoryRepository;

@Service
@Transactional
public class TaskHistoryService {

    @Autowired
    private TaskHistoryRepository taskHistoryRepository;

    /**
     * Record a task creation event
     */
    public void recordTaskCreation(Task task, User createdBy) {
        TaskHistory history = new TaskHistory(
            task, 
            createdBy, 
            HistoryAction.CREATE, 
            "task",
            null, 
            "Task created", 
            "New task created with initial data"
        );
        taskHistoryRepository.save(history);
    }

    /**
     * Record task field changes by comparing old and new values
     */
    public void recordTaskUpdate(Task oldTask, Task newTask, User modifiedBy) {
        // Record name changes
        if (!Objects.equals(oldTask.getName(), newTask.getName())) {
            recordFieldChange(newTask, modifiedBy, HistoryAction.UPDATE, "name", 
                            oldTask.getName(), newTask.getName());
        }

        // Record description changes
        if (!Objects.equals(oldTask.getDescription(), newTask.getDescription())) {
            recordFieldChange(newTask, modifiedBy, HistoryAction.UPDATE, "description", 
                            oldTask.getDescription(), newTask.getDescription());
        }

        // Record due date changes
        if (!Objects.equals(oldTask.getDueDate(), newTask.getDueDate())) {
            String oldDate = oldTask.getDueDate() != null ? oldTask.getDueDate().toString() : null;
            String newDate = newTask.getDueDate() != null ? newTask.getDueDate().toString() : null;
            recordFieldChange(newTask, modifiedBy, HistoryAction.UPDATE, "dueDate", oldDate, newDate);
        }

        // Record assignee changes
        if (!Objects.equals(getAssigneeId(oldTask), getAssigneeId(newTask))) {
            String oldAssignee = getAssigneeName(oldTask);
            String newAssignee = getAssigneeName(newTask);
            HistoryAction action = newTask.getAssignee() != null ? HistoryAction.ASSIGN : HistoryAction.UNASSIGN;
            recordFieldChange(newTask, modifiedBy, action, "assignee", oldAssignee, newAssignee);
        }

        // Record status changes
        if (!Objects.equals(getStatusId(oldTask), getStatusId(newTask))) {
            String oldStatus = getStatusName(oldTask);
            String newStatus = getStatusName(newTask);
            recordFieldChange(newTask, modifiedBy, HistoryAction.STATUS_CHANGE, "status", oldStatus, newStatus);
        }

        // Record priority changes
        if (!Objects.equals(getPriorityId(oldTask), getPriorityId(newTask))) {
            String oldPriority = getPriorityName(oldTask);
            String newPriority = getPriorityName(newTask);
            recordFieldChange(newTask, modifiedBy, HistoryAction.PRIORITY_CHANGE, "priority", oldPriority, newPriority);
        }
    }

    /**
     * Record task deletion
     */
    public void recordTaskDeletion(Task task, User deletedBy) {
        TaskHistory history = new TaskHistory(
            task, 
            deletedBy, 
            HistoryAction.DELETE, 
            "task",
            "Task existed", 
            null, 
            "Task deleted"
        );
        taskHistoryRepository.save(history);
    }

    /**
     * Get complete history for a task
     */
    @Transactional(readOnly = true)
    public List<TaskHistory> getTaskHistory(Long taskId) {
        return taskHistoryRepository.findByTaskIdOrderByTimestampDesc(taskId);
    }

    /**
     * Get paginated history for a task
     */
    @Transactional(readOnly = true)
    public Page<TaskHistory> getTaskHistory(Long taskId, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        return taskHistoryRepository.findByTaskIdOrderByTimestampDesc(taskId, pageable);
    }

    /**
     * Get project-wide task history
     */
    @Transactional(readOnly = true)
    public List<TaskHistory> getProjectHistory(Long projectId) {
        return taskHistoryRepository.findByProjectIdOrderByTimestampDesc(projectId);
    }

    /**
     * Get paginated project-wide task history
     */
    @Transactional(readOnly = true)
    public Page<TaskHistory> getProjectHistory(Long projectId, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        return taskHistoryRepository.findByProjectIdOrderByTimestampDesc(projectId, pageable);
    }

    /**
     * Get recent activity for a project (last 50 entries)
     */
    @Transactional(readOnly = true)
    public List<TaskHistory> getRecentProjectActivity(Long projectId) {
        Pageable pageable = PageRequest.of(0, 50);
        return taskHistoryRepository.findRecentHistoryByProjectId(projectId, pageable);
    }

    // Helper methods for safe field access

    private void recordFieldChange(Task task, User modifiedBy, HistoryAction action, 
                                 String fieldName, String oldValue, String newValue) {
        TaskHistory history = new TaskHistory(
            task, 
            modifiedBy, 
            action, 
            fieldName,
            oldValue, 
            newValue, 
            null
        );
        taskHistoryRepository.save(history);
    }

    private Long getAssigneeId(Task task) {
        return task.getAssignee() != null ? task.getAssignee().getId() : null;
    }

    private String getAssigneeName(Task task) {
        return task.getAssignee() != null ? task.getAssignee().getUsername() : null;
    }

    private Long getStatusId(Task task) {
        return task.getStatus() != null ? task.getStatus().getId() : null;
    }

    private String getStatusName(Task task) {
        return task.getStatus() != null ? task.getStatus().getName() : null;
    }

    private Long getPriorityId(Task task) {
        return task.getPriority() != null ? task.getPriority().getId() : null;
    }

    private String getPriorityName(Task task) {
        return task.getPriority() != null ? task.getPriority().getName() : null;
    }
}

File: apps/web/src/app/tasks/services/task-history/task-history.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { lastValueFrom } from 'rxjs';

export interface TaskHistory {
  id: number;
  task: {
    id: number;
    name: string;
  };
  modifiedBy: {
    id: number;
    username: string;
    email: string;
  };
  action: 'CREATE' | 'UPDATE' | 'DELETE' | 'ASSIGN' | 'UNASSIGN' | 'STATUS_CHANGE' | 'PRIORITY_CHANGE';
  fieldName: string;
  oldValue: string | null;
  newValue: string | null;
  timestamp: string;
  comment: string | null;
}

export interface TaskHistoryPage {
  content: TaskHistory[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
}

@Injectable({ providedIn: 'root' })
export class TaskHistoryService {
  private apiUrl = 'http://localhost:8080/api';
  private http = inject(HttpClient);

  /**
   * Get history for a specific task
   * @param taskId - The ID of the task
   * @param page - Page number (0-based)
   * @param size - Number of items per page
   */
  getTaskHistory = (taskId: number, page: number = 0, size: number = 20): Promise<TaskHistoryPage> => {
    const params = { page: page.toString(), size: size.toString() };
    return lastValueFrom(
      this.http.get<TaskHistoryPage>(`${this.apiUrl}/tasks/${taskId}/history`, { params })
    );
  };

  /**
   * Get all history for a specific task (no pagination)
   * @param taskId - The ID of the task
   */
  getAllTaskHistory = (taskId: number): Promise<TaskHistory[]> => {
    const params = { size: '0' }; // size=0 means no pagination
    return lastValueFrom(
      this.http.get<TaskHistory[]>(`${this.apiUrl}/tasks/${taskId}/history`, { params })
    );
  };

  /**
   * Get project-wide task history
   * @param projectId - The ID of the project
   * @param page - Page number (0-based)
   * @param size - Number of items per page
   */
  getProjectTaskHistory = (projectId: number, page: number = 0, size: number = 50): Promise<TaskHistoryPage> => {
    const params = { page: page.toString(), size: size.toString() };
    return lastValueFrom(
      this.http.get<TaskHistoryPage>(`${this.apiUrl}/projects/${projectId}/tasks/history`, { params })
    );
  };

  /**
   * Get recent project activity (last 50 entries)
   * @param projectId - The ID of the project
   */
  getRecentProjectActivity = (projectId: number): Promise<TaskHistory[]> => {
    return lastValueFrom(
      this.http.get<TaskHistory[]>(`${this.apiUrl}/projects/${projectId}/activity`)
    );
  };

  /**
   * Format action for display
   */
  formatAction(action: TaskHistory['action']): string {
    const actionMap = {
      'CREATE': 'Created',
      'UPDATE': 'Updated',
      'DELETE': 'Deleted',
      'ASSIGN': 'Assigned',
      'UNASSIGN': 'Unassigned',
      'STATUS_CHANGE': 'Status Changed',
      'PRIORITY_CHANGE': 'Priority Changed'
    };
    return actionMap[action] || action;
  }

  /**
   * Format field name for display
   */
  formatFieldName(fieldName: string): string {
    const fieldMap: Record<string, string> = {
      'name': 'Name',
      'description': 'Description',
      'dueDate': 'Due Date',
      'assignee': 'Assignee',
      'status': 'Status',
      'priority': 'Priority',
      'task': 'Task'
    };
    return fieldMap[fieldName] || fieldName;
  }

  /**
   * Get a summary of the change for display
   */
  getChangeSummary(historyItem: TaskHistory): string {
    const action = this.formatAction(historyItem.action);
    const field = this.formatFieldName(historyItem.fieldName);
    
    if (historyItem.action === 'CREATE') {
      return `${action} task`;
    }
    
    if (historyItem.action === 'DELETE') {
      return `${action} task`;
    }
    
    if (historyItem.action === 'ASSIGN' || historyItem.action === 'UNASSIGN') {
      return `${action} ${historyItem.newValue || 'task'}`;
    }
    
    if (historyItem.oldValue && historyItem.newValue) {
      return `${action} ${field} from "${historyItem.oldValue}" to "${historyItem.newValue}"`;
    }
    
    if (historyItem.newValue) {
      return `${action} ${field} to "${historyItem.newValue}"`;
    }
    
    if (historyItem.oldValue) {
      return `${action} ${field} from "${historyItem.oldValue}"`;
    }
    
    return `${action} ${field}`;
  }
}
